src/dev/adam_practice_tagfinder.py

```diff
@@ -0,0 +1,66 @@
+import bs4
+
+with open("sample_webpage.html") as html_file:
+    html = html_file.read()
+
+soup = bs4.BeautifulSoup(html, "html.parser")
+
+def p_tag_finder(my_soup):
+
+    # Lines 12 - 17 find all the <p> tags in the html file and count them, creating the...
+    # ...variable called 'total_number_of_p_tags'.
+    p_tags = my_soup.findAll('p')
+
+    total_number_of_p_tags = 0
+
+    for item in p_tags:
+        total_number_of_p_tags +=1
+
+    # Lines 21 - 24 find ALL parents of EVERY <p> tag and add them to the list...
+    # ...called 'p_tag_parents'.
+    p_tag_parents = []
+
+    for tag in p_tags:
+        p_tag_parents.append(tag.parents)
+
+    # Lines 30 - 40 iterate through this list of <p> tag parents and count...
+    # ...the number of <p> tags for each parent, creating a dictionary called 'p_tag_dict'...
+    # ...where the keys are the names of these parent tags and their (integer) values are...
+    # ...the number of <p> tags which they each contain.
+    sub_tag_p_number = 0
+
+    p_tag_dict = {}
+
+    for parent in p_tag_parents:
+        for sub_tag in parent:
+            sub_tag_contents = sub_tag.findAll('p')
+            for tag in sub_tag_contents:
+                sub_tag_p_number +=1
+                p_tag_dict.update({sub_tag.name: sub_tag_p_number})
```

Comment:
According to https://stackoverflow.com/questions/15456158/python-dict-update-vs-subscript-to-add-a-single-key-value-pair the following is a more efficient way of assigning key value pairs in dictionaries:

```suggestion
                p_tag_dict[sub_tag.name] = sub_tag_p_number
```
First of all as I said before you might be able to access the number if things in sub_tag_contents just by getting the len. Even if that doesn't work and you do need to loop to count them you will be updating this dictionary for every item in the loop which you don't want to do. so you should place this outside the loop.

Here you are looping over parents then looping over what is presumably the children of these parents, then counting the subtags of these children, how do you know this is enough layers to go down to find the true youngest ancestor?